### PR TITLE
cmd/kaf: Avro decoding support

### LINKED ISSUE #38: Integrate with Confluent schema registry (Avro)
Would be nice to have a non-JVM alternative to kafka-avro-console-consumer

### PR DESCRIPTION
**Pull request by tzetter**
This adds support for decoding Avro-encoded messages using a Confluent Schema Registry. If a schema registry URL is set an attempt is made to print out Avro-encoded message keys and values in JSON-form.

This resolves #38.

### CODE REVIEW

File: cmd/kaf/consume.go [hunk 1]
```diff
@@ -142,3 +154,21 @@ var consumeCmd = &cobra.Command{
 
 	},
 }
+
+func tryAvroDecode(b []byte) []byte {
```
[the remark this answers is not in the document]
Comment (tzetter) [reply cmd/kaf/consume.go]: You are correct. I'll add some error handling.

File: avro/schema.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		<-cc.done
+		return cc.codec, cc.err
+	}
+
+	// Create the cachedCodec with a promise of a future value.
+	cc = &cachedCodec{done: make(chan struct{})}
+	c.codecsBySchemaID[schemaID] = cc
+	c.mu.Unlock()
+
+	defer func() {
```
Comment (birdayz) [inline avro/schema.go]: to be honest i have trouble understanding the purpose of this construct, also with the channel and the error. why don't we just directly set the codec here instead of doing this in defer. can you elaborate?
also i think a normal mutex for this function is sufficient. i doubt we'll have performance issues.

Comment (mafredri) [reply avro/schema.go]: Apologies, this construct is my handiwork so I'll try to answer it instead of @tzetter.

The purpose of the defer is to assign the returned values to `cc` (`cachedCodec`) no matter which of the three code paths are taken by accessing the variables from the return signature (`(codec *goavro.Codec, err error)`). If we did not we would need to `cc.err = err; close(cc.done);` in both of the `if err != nil` check below. And in case of no error we would do `cc.codec = codec; close(cc.done);`. So the construct just ensures that we always finalize the `cachedCodec`.

I realize that is not easy to read in a sentence, so I'll illustrate how this construct would look without the defer:

```go
	schema, err := c.client.GetSchemaById(schemaID)
	if err != nil {
		cc.err = err
		close(cc.done)
		return nil, err
	}

 	codec, err = goavro.NewCodec(schema)
	if err != nil {
		cc.err = err
		close(cc.done)
		return nil, err
	}

	cc.codec = codec
	close(cc.done)
 	return codec, nil
```

The purpose of the done channel is to inform other goroutines that `cc.codec` and `cc.err` are ready to be read. Let's say one routine starts fetching the codec and has inserted its `&cachedCodec{done: make(chan struct{})}`, any other routine that tries to the same codec will end up waiting on `<-cc.done`. Thus, when it closes all routines can carry on and the values will be present. Kind of like a JavaScript promise, or similar construct.

> also i think a normal mutex for this function is sufficient. i doubt we'll have performance issues.

As in `sync.Mutex` and holding it for the entire duration? I suppose that would work as well, although it would force the other partition consumers to wait whenever there is an uncached codec. It's probably acceptable though.

Comment (birdayz) [reply avro/schema.go]: i guess it's a reasonable solution, even if the readability is not ideal. thanks for the explanation

**Comment by birdayz** [comment]
can you please rebase ? then we should be ready to merge.

**Comment by tzetter** [comment]
It has been rebased

**Comment by cricket007** [comment]
Question: Do both key and value have to be Avro for the full message to be printed? In other words, what if the _keys aren't_ Avro but just the values are? Is there a flag for configuring that? e.g. having `--key-deserializer` and `--value-deserializer` comparable?

**Comment by tzetter** [comment]
If a schema registry URL is set it will attempt to decode both the key and value. If either one isn't Avro-encoded it will be printed out in raw form. Therefore no extra configuration is needed. Checking if a message is avro-encoded is so fast that I don't see a reason to change this behavior, at least until we get support for other serializers (e.g. protobuf).

TL;DR you can mix and match however you like. It will detect which messages/keys are Avro.
